### PR TITLE
Add Export script

### PR DESCRIPTION
**Pull request by alessandropalla**
The export script generate OpenVINO and ONNX models of detection and classification NN
Then the OV classes in openvino_models.py provides a 1:1 mapping to the old torch classses for a plug&play experience

### CODE REVIEW

File: src/domain/openvino_models.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import openvino.properties as props
+import openvino as ov
+import numpy as np
+import torch
+
+core = ov.Core()
+core.set_property({props.cache_dir: "cache"})
+#TODO: Add NPU specific properties
+
+class DetectionModel():
```
Comment (cdimauro) [inline src/domain/openvino_models.py:11]: Add another blank line.

Comment (cdimauro) [inline src/domain/openvino_models.py:11]: Redundant ()

File: src/domain/openvino_models.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    def __init__(self, device="cpu") -> None:
+        print("Loading Openvino model")
+        
+        ov_model = core.read_model("detection_model.onnx")
+        self.model = ov.compile_model(ov_model, device_name=device.upper())
+
+    def __call__(self, img: torch.tensor):
+       results = [torch.tensor(t) for t in self.model(img).values()]
```
Comment (cdimauro) [inline src/domain/openvino_models.py:19]: It could be:

       return [torch.tensor(t) for t in self.model(img).values()]

However, is it really needed to give back a list? If not strictly required, an iterator could be a more efficient solution (for example, because the values are just used in a for loop):

       return (torch.tensor(t) for t in self.model(img).values())

File: src/domain/openvino_models.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+class OVMegaDetectorV5(pw_detection.MegaDetectorV5):
+    def __init__(self, device="cpu"):
+        self.model = DetectionModel(device)
+        self.device = "cpu"  # torch device, keep to CPU when using with OpenVINO
+
+
+class OVClassificationModel():
+    def __init__(self, weight_path, device):
+
```
Comment (cdimauro) [inline src/domain/openvino_models.py:31]: Redundant blank line

File: src/domain/openvino_models.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        :param test_loader: test dataloader: torch.utils.data.DataLoader
+        :return: numpy array of predictions without soft max
+        """
+        total_output = []
+        with torch.no_grad():
+            x = data.to(self.device) # ADJUSTMENT 2
+            if withsoftmax:
+                output = self.forward(x).softmax(dim=1)
+            else:
+                output = self.forward(x)
```
Comment (cdimauro) [inline src/domain/openvino_models.py:54]: output = self.forward(x).softmax(dim=1) if withsoftmax else self.forward(x)

or:
            output = self.forward(x)
            if withsoftmax:
                output = output.softmax(dim=1)

File: src/domain/openvino_models.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            total_output += output.tolist()
+
+        return np.array(total_output)
+
+    def loadWeights(self, path):
```
Comment (cdimauro) [inline src/domain/openvino_models.py:59]: loadWeights -> load_weights

File: src/gradio_demo.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
+from domain.ai_model import AiModel
+import gradio as gr
```
Comment (cdimauro) [inline src/gradio_demo.py:2]: import gradio as gr

from domain.ai_model import AiModel

File: src/gradio_demo.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+ai_model = AiModel()
+
+
+def detection_from_url(url):
+    det_data = ai_model.process_image_from_url(url, "test_image", True)
+    img_path, det_results, detected_img_path = det_data
+    return img_path, det_results
+
```
Comment (cdimauro) [inline src/gradio_demo.py:11]: Add blank line

File: src/gradio_demo.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        print("Running classification on detection result(s)...")
+        img_path, classified_animals = ai_model.classify(img_path, det_results)
+        str = f"Found {len(classified_animals)} animal{'s' if len(classified_animals) > 1 else ''}:\n"
+        for animals in classified_animals:
+            animal, confidence = animals['classification']
+            str += f"\t - Class: {animal}, confidence: {confidence*100:.2f}%\n"
+        return str
+    else:
+        return "No animal found"
+
```
Comment (cdimauro) [inline src/gradio_demo.py:23]: Add blank line

File: src/gradio_demo.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    return img_path, det_results
+
+def classification_from_path(img_path, det_results):
+    if len(det_results["detections"].xyxy) > 0:
+        print("Running classification on detection result(s)...")
+        img_path, classified_animals = ai_model.classify(img_path, det_results)
+        str = f"Found {len(classified_animals)} animal{'s' if len(classified_animals) > 1 else ''}:\n"
+        for animals in classified_animals:
+            animal, confidence = animals['classification']
+            str += f"\t - Class: {animal}, confidence: {confidence*100:.2f}%\n"
```
Comment (cdimauro) [inline src/gradio_demo.py:19]: str += is not recommended if the list could be huge. In those cases it's better to collect all strings into a list, and then use ''.join() to build the final string.

File: src/gradio_demo.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+def process_url(url):
+    # 1: Detection
+    detection_path, det_results = detection_from_url(url)
+    
+    # 2: Classification
+    classification_result = classification_from_path(detection_path, det_results)
+    return detection_path, classification_result
+
+url = "https://www.provincia.bz.it/agricoltura-foreste/fauna-caccia-pesca/images/braunbaer_6016_L.jpg"
```
Comment (cdimauro) [inline src/gradio_demo.py:32]: CONSTANTS at the top of the module.

File: src/requirements.txt [hunk 11]
```diff
@@ -1,12 +1,14 @@
...
+opencv-python
+timm
+validators
+torch
+torchvision
+torchaudio
+intel-npu-acceleration-library
+PytorchWildlife
+openvino
+onnx
```
Comment (cdimauro) [inline src/requirements.txt:14]: cv2_enumerate_cameras
enum
ipykernel
intel-npu-acceleration-library
keyring
onnx
opencv-python
openvino
PytorchWildlife
timm
torch
torchvision
torchaudio
validators

File: src/domain/ai_model.py [hunk 12]
```diff
@@ -7,9 +7,9 @@
 import cv2
 from PIL import Image
 import numpy as np
-from PytorchWildlife.models import detection as pw_detection
 from PytorchWildlife.data import transforms as pw_trans
 #import intel_npu_acceleration_library as npu_lib
+from domain.openvino_models import OVMegaDetectorV5
 from PytorchWildlife import utils as pw_utils
 from domain.classify_detections import Classifier, txt_animalclasses
```
Comment (cdimauro) [inline src/domain/ai_model.py:14]: import cv2
import numpy as np
from PIL import Image
from PytorchWildlife import utils as pw_utils
from PytorchWildlife.data import transforms as pw_trans

from domain.classify_detections import Classifier, txt_animalclasses
from domain.openvino_models import OVMegaDetectorV5

File: src/domain/classify_detections.py [hunk 13]
```diff
@@ -13,6 +13,7 @@
 from torch import tensor
 import torch.nn as nn
 from torchvision.transforms import InterpolationMode, transforms
+from domain.openvino_models import OVClassificationModel
```
Comment (cdimauro) [inline src/domain/classify_detections.py:16]: import sys

import numpy as np
import timm
import torch
from torch import tensor
import torch.nn as nn
from torchvision.transforms import InterpolationMode, transforms

from domain.openvino_models import OVClassificationModel

File: src/domain/classify_detections.py [hunk 14]
```diff
@@ -80,7 +81,7 @@
```
Comment (cdimauro) [inline src/domain/classify_detections.py:81]: Add blank line

File: src/domain/openvino_models.py [hunk 15]
```diff
@@ -0,0 +1,57 @@
+from PytorchWildlife.models import detection as pw_detection
+import openvino as ov
+import numpy as np
+import torch
```
Comment (cdimauro) [inline src/domain/openvino_models.py]: import numpy as np
import openvino as ov
from PytorchWildlife.models import detection as pw_detection
import torch

**Review by cdimauro — COMMENTED** [review]
No further comments: already gave on previous PR.

File: src/domain/openvino_models.py [hunk 16]
```diff
@@ -1,13 +1,17 @@
 from PytorchWildlife.models import detection as pw_detection
+import openvino.properties as props
 import openvino as ov
 import numpy as np
 import torch
```
Comment (cdimauro) [inline src/domain/openvino_models.py]: import numpy as np
import openvino as ov
import openvino.properties as props
from PytorchWildlife.models import detection as pw_detection
import torch

File: src/gradio_demo.py [hunk 17]
```diff
@@ -0,0 +1,54 @@
+
+import sys
+import os
+
+sys.path.append("../src")
+from domain.ai_model import AiModel
+import gradio as gr
```
Comment (cdimauro) [inline src/gradio_demo.py]: import os
import sys

import gradio as gr

sys.path.append("../src")
from domain.ai_model import AiModel

File: src/gradio_demo.py [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+    detection_path, det_results = detection_from_url(url)
+    
+    # 2: Classification
+    classification_result = classification_from_path(detection_path, det_results)
+    return detection_path, classification_result
+
+url = "https://www.provincia.bz.it/agricoltura-foreste/fauna-caccia-pesca/images/braunbaer_6016_L.jpg"
+
+
+# Configurazione dell'interfaccia Gradio
```
Comment (cdimauro) [inline src/gradio_demo.py]: English

File: src/gradio_demo.py [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+    inputs=gr.Textbox(label="Image URL"),
+    outputs=[
+        gr.Image(label="Detection Results"),
+        gr.Textbox(label="Classification Results")
+    ],
+    title="WADAS Demo",
+    description="Enter a URL to get the detection and classification"
+)
+
+# Esegui l'interfaccia
```
Comment (cdimauro) [inline src/gradio_demo.py]: Redundant

File: src/domain/ai_model.py (hunk 12 above)
Comment (stefanodellosa-personal) [reply src/domain/ai_model.py:14]: @alessandropalla removing from "_PytorchWildlife.models import detection as pw_detection_" won't affect WADS to work in non-intel-npu scenario?

File: src/domain/classify_detections.py (hunk 14 above)
Comment (stefanodellosa-personal) [reply src/domain/classify_detections.py:81]: I guess the comment is to add additional blankline before class definition..

File: src/domain/openvino_models.py [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+        self.model = ov.compile_model(ov_model, device_name=device.upper())
+        self.device = "cpu"
+
+    def forward(self, input):
+        results = [torch.tensor(t) for t in self.model(input).values()]
+        if len(results) == 1:
+            return results[0]
+        return results
+
+    def predict(self, data, withsoftmax=True):
```
Comment (stefanodellosa-personal) [inline src/domain/openvino_models.py:42]: Is this copy pasted code?

File: src/domain/classify_detections.py (hunk 14 above)
Comment (cdimauro) [reply src/domain/classify_detections.py:82]: Exactly! :)

File: demo/gradio_demo.py [hunk 21]
```diff
@@ -1,22 +1,25 @@
-from domain.ai_model import AiModel
 import gradio as gr
+from src.domain.ai_model import AiModel
 
 ai_model = AiModel()
-
+url = "https://www.provincia.bz.it/agricoltura-foreste/fauna-caccia-pesca/images/braunbaer_6016_L.jpg"
```
Comment (cdimauro) [inline demo/gradio_demo.py:5]: URL = "https://www.provincia.bz.it/agricoltura-foreste/fauna-caccia-pesca/images/braunbaer_6016_L.jpg"

File: demo/gradio_demo.py [hunk 22]
```diff
@@ -1,22 +1,25 @@
...
 
 def detection_from_url(url):
```
Comment (cdimauro) [inline demo/gradio_demo.py:7]: Add blank line.

File: src/domain/ai_model.py [hunk 23]
```diff
@@ -5,13 +5,13 @@
...
 import cv2
-from PIL import Image
 import numpy as np
-from PytorchWildlife.data import transforms as pw_trans
-#import intel_npu_acceleration_library as npu_lib
-from domain.openvino_models import OVMegaDetectorV5
+from PIL import Image
 from PytorchWildlife import utils as pw_utils
-from domain.classify_detections import Classifier, txt_animalclasses
+from PytorchWildlife.data import transforms as pw_trans
```
Comment (cdimauro) [inline src/domain/ai_model.py]: import datetime
import logging
import os

import cv2
import numpy as np
from PIL import Image
from PytorchWildlife import utils as pw_utils
from PytorchWildlife.data import transforms as pw_trans
import requests

File: model/export_models.py [hunk 24]
```diff
@@ -1,32 +1,33 @@
-from PytorchWildlife.models import detection as pw_detection
-from PytorchWildlife.data import transforms as pw_trans
-from torchvision.transforms import InterpolationMode, transforms
-from torch.export import export
-import argparse
-from PIL import Image
+import sys
+import os
```
Comment (cdimauro) [inline model/export_models.py:2]: import os
import sys

File: src/domain/ai_model.py (hunk 12 above)
Comment (alessandropalla) [reply src/domain/ai_model.py]: No, as long as we use onnx for inference (and imho we should)

**Comment by alessandropalla** [comment]
@cdimauro I think they were handled by @stefanodellosa-personal  [here](https://github.com/stefanodellosa-personal/WADAS/pull/13/commits/d3298841b57761aaf2e47d1db3a897d2d6cd1720..019cfd7eeb5f0442515b9c7f5610bf58b8f84d41). I haven't fully checked that everything was handled tho. However I was thinking to have a pre-commit style check as implemented [here](https://github.com/stefanodellosa-personal/WADAS/pull/15). Feel free to add linting rules that you like the most. My idea was to merge that PR, update this branch with the changes, do an automatic style fix and push it back what do you think?

**Comment by cdimauro** [comment]
> @cdimauro I think they were handled by @stefanodellosa-personal [here](https://github.com/stefanodellosa-personal/WADAS/pull/13/commits/d3298841b57761aaf2e47d1db3a897d2d6cd1720..019cfd7eeb5f0442515b9c7f5610bf58b8f84d41). I haven't fully checked that everything was handled tho. However I was thinking to have a pre-commit style check as implemented [here](https://github.com/stefanodellosa-personal/WADAS/pull/15). Feel free to add linting rules that you like the most. My idea was to merge that PR, update this branch with the changes, do an automatic style fix and push it back what do you think?

I think that it's the way to go. :) Better delegate to machines/tools doing such work, and let human beings focus on more interesting stuff. ;)
Your PR is huge. It'll take same time to review it. Today I hardly have time, but let's see if I can do something at least.

**Comment by alessandropalla** [comment]
@cdimauro Thanks take your time :) The other PR is mostly style fix of the entire repo. The only meaningful piece is the `.pre-commit-config.yaml` file that has the linting rule. All the other changes are done automatically by the tool.

File: src/domain/openvino_models.py (hunk 4 above)
Comment (stefanodellosa-personal) [reply src/domain/openvino_models.py:57]: @alessandropalla do you want to handle this?

File: src/domain/openvino_models.py (hunk 2 above)
Comment (stefanodellosa-personal) [reply src/domain/openvino_models.py]: @alessandropalla do you want to handle this?

File: model/export_models.py
Comment (stefanodellosa-personal) [inline model/export_models.py:1]: It would be great to add README in "model" folder to teach users about
1- What is this script for
2- How to run it
3- Where to place generated files

File: src/domain/openvino_models.py [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+import openvino.properties as props
+from PytorchWildlife.models import detection as pw_detection
+import torch
+
+core = ov.Core()
+core.set_property({props.cache_dir: "cache"})
+# TODO: Add NPU specific properties
+
+
+class DetectionModel:
```
Comment (stefanodellosa-personal) [inline src/domain/openvino_models.py:14]: Add docustring comments in class and methods

File: src/domain/openvino_models.py [hunk 26]
```diff
@@ -0,0 +1,64 @@
...
+import openvino as ov
+import openvino.properties as props
+from PytorchWildlife.models import detection as pw_detection
+import torch
+
+core = ov.Core()
+core.set_property({props.cache_dir: "cache"})
+# TODO: Add NPU specific properties
+
+
```
Comment (stefanodellosa-personal) [inline src/domain/openvino_models.py:13]: Add logger like other classes, so that we can use it and embedd it in UI (or daemon later). Plus, consider logging debug or info messages to have feedback on progress

**Comment by stefanodellosa-personal** [comment]
> The export script generate OpenVINO and ONNX models of detection and classification NN Then the OV classes in openvino_models.py provides a 1:1 mapping to the old torch classses for a plug&play experience

Performance improvement is massive! Great Ale!

@alessandropalla  please add in src/README instruction about preconditions, i.e., generation and copy of model files into WADAS root folder.

File: .gitignore [hunk 27]
```diff
@@ -1,62 +1,66 @@
...
+bower_components/
+
+# Images directories
+classification_output/
+detection_output/
+url_imgs/
+debug/
+wadas_motion_detection/
+demo_output/
+
```
Comment (stefanodellosa-personal) [inline .gitignore:55]: Pristine (or add, i don't see it in previous version but was supposed to be there) the "log" folder, where all logs files are stored. E.g., WADAS main log, FTP server log, etc..

**Review by stefanodellosa-personal — CHANGES_REQUESTED** [review]
Please see comments in files and conversation.
